src/VecSim/spaces/IP/IP_AVX2_BF16.h

```diff
@@ -0,0 +1,132 @@
+    // convert next  0:3, 8:11 bf16 to 8 floats
+    __m256i bf16_low1 = _mm256_unpacklo_epi16(zeros, v1); // AVX2
+    __m256i bf16_low2 = _mm256_unpacklo_epi16(zeros, v2);
+
+    // compute dist
+    sum_prod = _mm256_add_ps(
+        sum_prod, _mm256_mul_ps(_mm256_castsi256_ps(bf16_low1), _mm256_castsi256_ps(bf16_low2)));
+}
+
+static inline void InnerProductHighHalfStep(__m256i v1, __m256i v2, __m256i zeros,
+                                            __m256 &sum_prod) {
+    // convert next 4:7, 12:15 bf16 to 8 floats
+    __m256i bf16_high1 = _mm256_unpackhi_epi16(zeros, v1);
+    __m256i bf16_high2 = _mm256_unpackhi_epi16(zeros, v2);
+
+    // compute dist
+    sum_prod = _mm256_add_ps(
+        sum_prod, _mm256_mul_ps(_mm256_castsi256_ps(bf16_high1), _mm256_castsi256_ps(bf16_high2)));
+}
+
+static inline void InnerProductStep(bfloat16 *&pVect1, bfloat16 *&pVect2, __m256 &sum_prod) {
+    // load 16 bf16 elements
+    __m256i v1 = _mm256_lddqu_si256((__m256i *)pVect1); // avx
+    pVect1 += 16;
+    __m256i v2 = _mm256_lddqu_si256((__m256i *)pVect2);
+    pVect2 += 16;
+
+    __m256i zeros = _mm256_setzero_si256(); // avx
+
+    // Compute dist for 0:3, 8:11 bf16
+    InnerProductLowHalfStep(v1, v2, zeros, sum_prod);
+
+    // Compute dist for 4:7, 12:15 bf16
+    InnerProductHighHalfStep(v1, v2, zeros, sum_prod);
+}
+
+template <unsigned char residual> // 0..31
+float BF16_InnerProductSIMD32_AVX2(const void *pVect1v, const void *pVect2v, size_t dimension) {
+    // cast to bfloat16 *
+    bfloat16 *pVect1 = (bfloat16 *)pVect1v;
+    bfloat16 *pVect2 = (bfloat16 *)pVect2v;
+
+    // define end
+    const bfloat16 *pEnd1 = pVect1 + dimension;
+
+    // declare sum
+    __m256 sum_prod = _mm256_setzero_ps();
+
+    // handle first residual % 16 elements
+    if (residual % 16) {
+        // load all 16 elements to a 256 bit register
+        __m256i v1 = _mm256_lddqu_si256((__m256i *)pVect1); // avx
+        pVect1 += residual % 16;
+        __m256i v2 = _mm256_lddqu_si256((__m256i *)pVect2);
+        pVect2 += residual % 16;
+
+        // unpack  0:3, 8:11 bf16 to 8 floats
+        __m256i zeros = _mm256_setzero_si256();
+        __m256i v1_low = _mm256_unpacklo_epi16(zeros, v1);
+        __m256i v2_low = _mm256_unpacklo_epi16(zeros, v2);
+
+        __m256 low_mul = _mm256_mul_ps(_mm256_castsi256_ps(v1_low), _mm256_castsi256_ps(v2_low));
+        if (residual % 16 <= 4) {
+            const unsigned char elem_to_calc = residual % 16;
+            const __mmask8 mask = (1 << elem_to_calc) - 1;
+            low_mul = _mm256_blend_ps(_mm256_setzero_ps(), low_mul, mask);
+        } else {
+            __m256i v1_high = _mm256_unpackhi_epi16(zeros, v1);
+            __m256i v2_high = _mm256_unpackhi_epi16(zeros, v2);
+            __m256 high_mul =
+                _mm256_mul_ps(_mm256_castsi256_ps(v1_high), _mm256_castsi256_ps(v2_high));
+            if (4 < residual % 16 && residual % 16 <= 8) {
```

Comment:
This one was essential to allow declaring all `elem_to_calc` and `mask` as `constexpr` (otherwise the compiler 
threw an error that the shift exceeds maximum unsigned char)